accel/tcg/log_instr.c

```diff
@@ -0,0 +1,513 @@
+#define log_assert(x) assert((x))
+#else
+#define log_assert(x)
+#endif
+
+struct log_reginfo {
+    uint16_t flags;
+#define LRI_CAP_REG    1
+#define LRI_HOLDS_CAP  2
+
+    const char *name;
+    union {
+        target_ulong gpr;
+#ifdef TARGET_CHERI
+        cap_register_t cap;
+#endif
+    };
+};
+typedef struct log_reginfo log_reginfo_t;
+
+#define reginfo_is_cap(ri) (ri->flags & LRI_CAP_REG)
+#define reginfo_has_cap(ri) (reginfo_is_cap(ri) && (ri->flags & LRI_HOLDS_CAP))
+
+extern int cl_default_trace_format;
+
+static cpu_log_instr_info_t *cpu_get_log(CPUState *cpu)
+{
+    return &cpu->log_info;
+}
+
+static cpu_log_instr_info_t *target_cpu_get_log(CPUArchState *env)
+{
+    return &env_cpu(env)->log_info;
+}
+
+/*
+ * Emit textual trace entry to the log.
+ */
+static void emit_text_trace(CPUArchState *env, cpu_log_instr_info_t *log)
+{
+    int i;
+
+    /* Dump main instruction log */
+#if defined(TARGET_RISCV) && defined(CONFIG_RVFI_DII)
+    /*
+     * TODO(am2419): what to do with injected instructions?
+     * Is the rvfi_dii_trace state valid at log commit?
+     */
+    if (env->rvfi_dii_have_injected_insn) {
+        uint32_t insn = env->rvfi_dii_trace.rvfi_dii_insn;
+        target_disas_buf(stderr, env_cpu(env), &insn, sizeof(insn),
+                         PC_ADDR(env), 1);
+    } else
+#endif
+    {
+        log_target_disas(env_cpu(env), log->pc, /*only one*/-1);
+    }
+
+    /* Dump memory access */
+
+    /* Dump register changes and side-effects */
+    for (i = 0; i < log->regs->len; i++) {
+        log_reginfo_t *rinfo = &g_array_index(log->regs, log_reginfo_t, i);
+
+#ifndef TARGET_CHERI
+        log_assert(!reginfo_is_cap(rinfo) && "Register marked as capability "
+                   "register whitout CHERI support");
+#else
+        if (reginfo_is_cap(rinfo)) {
+            if (reginfo_has_cap(rinfo))
+                qemu_log("newapi    Write %s|" PRINT_CAP_FMTSTR_L1 "\n"
+                         "             |" PRINT_CAP_FMTSTR_L2 "\n",
+                         rinfo->name,
+                         PRINT_CAP_ARGS_L1(&rinfo->cap),
+                         PRINT_CAP_ARGS_L2(&rinfo->cap));
+            else
+                qemu_log("newapi  %s <- " TARGET_FMT_lx
+                         " (setting integer value)\n",
+                         rinfo->name, rinfo->gpr);
+        } else
+#endif
+        {
+            qemu_log("newapi    Write %s = " TARGET_FMT_lx "\n", rinfo->name,
+                     rinfo->gpr);
+        }
+    }
+
+    /* Dump extra logged messages, if any */
+    if (log->txt_buffer->len > 0)
+        qemu_log("%s", log->txt_buffer->str);
+}
+
+/*
+ * Emit binary trace entry to the log.
+ */
+static void emit_cvtrace(cpu_log_instr_info_t *log)
+{
+    // TODO(am2419): how do we get the opcode from pc? this is a machine-dependant
+    // thing, maybe we can reuse disas or have an hook to determine the opcode length
+    // can we get it from next instruction pc?
```

Comment:
You should be able to lookup the next pc from the saved TCG metadata. However, I'm not sure if that works when we reach the end of the translation block.

Reply:
Thanks for confirming this, I will experiment a bit as I'm not sure yet.